test/test_prims.py

```diff
@@ -239,5 +240,31 @@ def test_mul_complex(self):
 
 instantiate_device_type_tests(TestPrims, globals())
 
+
+class TestRefs(TestCase):
+    @dtypes(torch.float32)
+    def test_constant_pad_nd_memory_format(self, device, dtype):
+        # Test memory format is preserved in unambiguous cases
+        for mf, ndim in (
+                (torch.channels_last, 4),
+                (torch.contiguous_format, 4),
+                (torch.channels_last_3d, 5),
+                (torch.contiguous_format, 5),
+        ):
+            a = torch.zeros([2] * ndim).to(memory_format=mf)
+            res = refs.constant_pad_nd(a, pad=[1] * (2 * ndim))
+            self.assertTrue(res.is_contiguous(memory_format=mf))
+
+        # Test that ambiguous cases default to contiguous
```

Comment:
this is not an ambiguous case (it's strides indicate that it is indeed channels-last, `suggest_memory_format` uses values of strides at sizes=1 to determine that), and in eager it would result in channels-last output:
```
In [3]:         a = torch.empty_strided((2, 1, 2, 2), stride=(4, 1, 2, 1))
   ...:         assert a.is_contiguous(memory_format=torch.channels_last)

In [4]: res = torch.constant_pad_nd(a, pad=[1] * 8)

In [5]: res.stride()
Out[5]: (48, 1, 12, 3)
```
ambiguous case would have contiguous-like strides, and would result in contiguous output:
```
In [6]: a = torch.empty_strided((2, 1, 2, 2), stride=(4, 4, 2, 1))

In [7]: a.is_contiguous(memory_format=torch.channels_last)
Out[7]: True

In [8]: res = torch.constant_pad_nd(a, pad=[1] * 8)

In [9]: res.stride()
Out[9]: (48, 16, 4, 1)
```

Reply:
I've added both cases to the test with a comparison against eager mode to be sure I'm not getting it wrong. In hindsight I got mixed up between `is_channels_last_` and `is_channels_last_contiguous_` which is less strict.